MericCelik/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,266 @@
+import java.util.ArrayList;
+
+/**Implementation of binary tree
+ * @author AycaMericCelik
+ * */
+
+public class BinaryTree<T> {
+    private BinaryTreeNode<T> root;
+
+
+    public BinaryTree(BinaryTreeNode<T> root){
+        this.root = root;
+    }
+
+    public BinaryTree(T root){
+        this(new BinaryTreeNode<>(root));
+    }
+
+    public BinaryTreeNode<T> getRoot() {
+        return root;
+    }
+
+    public void setRoot(BinaryTreeNode<T> root) {
+        this.root = root;
+    }
+
+
+    /**Finds the returns the node with given element if it occurs in the tree. Otherwise returns null. It does depth-first search in
+     * binary tree.
+     * Complexities: Average O(N), Worst O(N)(When we traverse all nodes),
+     * Best O(1)(When the root has the given element) --- N is the node count of the tree.
+     * @param root: root of the tree/subtree
+     * @param element: element that we are looking for
+     * @return the node with given element, null if not found.
+     * */
+    private BinaryTreeNode<T> findNodeWithDFS(BinaryTreeNode<T> root, T element){
+        if(root==null)
+            return null;
+        if(root.getKey().equals(element))
+            return root;
+        BinaryTreeNode<T> left = findNodeWithDFS(root.getLeftchild(), element);
```

Comment:
optional: as an optimisation you can check the nullness of the 'left' node here. If (left != null) return left and then there is no need to do a search on the right child. :) otherwise check the right child and just return that value:

```
BinaryTreeNode<T> left = findNodeWithDFS(root.getRightchild(), element);
if (left != null) {
    return left;
}
return findNodeWithDFS(root.getRightchild(), element);
```

Reply:
Oh, I see! Fixed, thank you! :+1: